File: design_patterns/C++/Behavioral/template_method.cxx
#include <iostream>

class Builder {
  public:
	// template method
	void build() {
		this->test();
		this->lint();
		this->assemble();
		this->deploy();
	}

	virtual void test() = 0;
	virtual void lint() = 0;
	virtual void assemble() = 0;
	virtual void deploy() = 0;
};

class AndroidBuilder: public Builder {
  public:
	void test() {
		std::cout << "Running Android tests" << std::endl;
	}

	void lint() {
		std::cout << "Linting the Android code" << std::endl;
	}

	void assemble() {
		std::cout << "Assembling the Android build" << std::endl;
	}

	void deploy() {
		std::cout << "Deploying Android build to server" << std::endl;
	}
};

class IosBuilder: public Builder {
  public:
	void test() {
		std::cout << "Running iOS tests" << std::endl;
	}

	void lint() {
		std::cout << "Linting the iOS code" << std::endl;
	}

	void assemble() {
		std::cout << "Assembling the iOS build" << std::endl;
	}

	void deploy() {
		std::cout << "Deploying iOS build to server" << std::endl;
	}
};

int main() {
	auto androidBuilder = new AndroidBuilder();
	androidBuilder->build();
	// =================================
	// Running Android tests
	// Linting the Android code
	// Assembling the Android build
	// Deploying Android build to server
	// =================================
	auto iosBuilder = new IosBuilder();
	iosBuilder->build();
	// =============================
	// Running iOS tests
	// Linting the iOS code
	// Assembling the iOS build
	// Deploying iOS build to server
	// =============================
	return 0;
}

File: design_patterns/Python/Behavioral/template_method.py
class Builder:

    def build(self):
        self.test()
        self.lint()
        self.assemble()
        self.deploy()

    def test(self):
        raise NotImplementedError

    def lint(self):
        raise NotImplementedError

    def assemble(self):
        raise NotImplementedError

    def deploy(self):
        raise NotImplementedError


class AndroidBuilder(Builder):

    def test(self):
        print('Running Android tests')

    def lint(self):
        print('Linting the Android code')

    def assemble(self):
        print('Assembling the Android build')

    def deploy(self):
        print('Deploying Android build to server')


class IosBuilder(Builder):

    def test(self):
        print('Running iOS tests')

    def lint(self):
        print('Linting the iOS code')

    def assemble(self):
        print('Assembling the iOS build')

    def deploy(self):
        print('Deploying iOS build to server')


if __name__ == '__main__':

    android_builder = AndroidBuilder()
    android_builder.build()
    # output:
    # Running Android tests
    # Linting the Android code
    # Assembling the Android build
    # Deploying Android build to server

    ios_builder = IosBuilder()
    ios_builder.build()
    # output:
    # Running iOS tests
    # Linting the iOS code
    # Assembling the iOS build
    # Deploying iOS build to server
